File: src/days/day2.rs
use std::cmp::{max, min};
use std::path::Path;

fn validate_report(r: &[i32]) -> bool {
    let pairs = r.iter().zip(r.iter().skip(1));
    let diffs: Vec<i32> = pairs.map(|(x, y)| x - y).collect();

    let min_diff = *diffs.iter().min().unwrap();
    let max_diff = *diffs.iter().max().unwrap();

    // False if they have different signs
    if min_diff * max_diff < 0 {
        return false;
    };

    let (abs_min, abs_max) = (
        min(min_diff.abs(), max_diff.abs()),
        max(min_diff.abs(), max_diff.abs()),
    );

    // False if the min abs diff is too small
    if abs_min < 1 {
        return false;
    };

    // False if the max abs diff is too big
    if abs_max > 3 {
        return false;
    };

    true
}

pub fn part1(data_path: &Path) -> u32 {
    // Read in the entire file
    let text = std::fs::read_to_string(data_path).unwrap();

    let reports: Vec<Vec<i32>> = text
        .lines()
        .map(|l| l.split_whitespace().map(|n| n.parse().unwrap()).collect())
        .collect();

    let num_safe_reports = reports
        .iter()
        .map(|r| if validate_report(r) { 1 } else { 0 })
        .sum();

    num_safe_reports
}

pub fn part2(data_path: &Path) -> u32 {
    let text = std::fs::read_to_string(data_path).unwrap();

    let reports: Vec<Vec<i32>> = text
        .lines()
        .map(|l| l.split_whitespace().map(|n| n.parse().unwrap()).collect())
        .collect();

    let mut num_safe_reports = 0;

    // Same as before, but try dropping each element
    for r in reports {
        let mut valid = validate_report(&r);

        for i in 0..r.len() {
            if valid {
                break;
            }
            let mut sub_r = r.clone();
            sub_r.remove(i); //Change to a filter, this is very expensive!
            valid = validate_report(&sub_r);
        }

        num_safe_reports += if valid { 1 } else { 0 };
    }

    num_safe_reports
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs::File;
    use std::io::Write;
    use std::path::PathBuf;
    use tempfile::tempdir;

    fn create_test_file() -> (tempfile::TempDir, File, PathBuf) {
        let test_input = "\
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9\n";

        let temp_dir = tempdir().unwrap();
        let f_path = temp_dir.path().join("test_input.txt");
        let mut temp_file = File::create(f_path.clone()).unwrap();
        write!(temp_file, "{}", test_input).unwrap();

        // have to return dir and file so they don't go out of scope
        (temp_dir, temp_file, f_path)
    }

    #[test]
    fn test_part1() {
        let (_d, _f, test_path) = create_test_file();
        let result = part1(&test_path);
        assert_eq!(result, 2);
    }

    #[test]
    fn test_part2() {
        let (_d, _f, test_path) = create_test_file();
        let result = part2(&test_path);
        assert_eq!(result, 4);
    }
}

File: python/advent2024/day2.py
import pathlib

def validate_report(report: list[str]) -> bool:
    nums = [int(n) for n in report]
    diffs = [l-r for (l,r) in zip(nums,nums[1:])]

    # Invalid if some diffs are negative and some are positive
    min_diff,max_diff = min(diffs),max(diffs)
    if min_diff*max_diff < 0:
        return False
    
    # Invalid if the absolute value of the min diff is < 1 or abs value of max diff is > 3
    abs_min,abs_max = min(abs(min_diff),abs(max_diff)),max(abs(min_diff),abs(max_diff))
    if abs_min < 1 or abs_max > 3:
        return False
    
    return True

def part1(f_path: pathlib.Path) -> int:
    with open(f_path) as f:
        num_valid_reports = sum([validate_report(l.strip().split()) for l in f])

    return num_valid_reports

def part2(f_path: pathlib.Path) -> int:
    num_valid_reports = 0

    with open(f_path) as f:
        for l in f:
            report = l.strip().split()
            valid = validate_report(report)
            for i in range(len(report)):
                if valid:
                    break

                valid = validate_report(report[:i] + report[i+1:])

            num_valid_reports += valid


    return num_valid_reports